StringIndexOutOfBoundsException: length=49; index=50(MsgBlockParser.detectBlockNext)

### Where? 
CreateMessageFragment -> fetch and update info about contacts

### Need to look at

```
Caused by: java.lang.StringIndexOutOfBoundsException: length=49; index=50
        at java.lang.String.substring(String.java:1971)
        at com.flowcrypt.email.example.file.MsgBlockParser.detectBlockNext(MsgBlockParser.kt:37)
```
### Full stack

```
[SUMMARY]
com.flowcrypt.email.util.exception.ManualHandledException: Handled manually:
java.lang.StringIndexOutOfBoundsException: length=49; index=50


[STACK_TRACE]
com.flowcrypt.email.util.exception.ManualHandledException: Handled manually:
        at com.flowcrypt.email.util.exception.ExceptionUtil$Companion.handleError(ExceptionUtil.kt:30)
        at com.flowcrypt.email.jetpack.viewmodel.ContactsViewModel$fetchAndUpdateInfoAboutContacts$1.invokeSuspend(ContactsViewModel.kt:171)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6558)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:469)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826)
Caused by: java.lang.StringIndexOutOfBoundsException: length=49; index=50
        at java.lang.String.substring(String.java:1971)
        at com.flowcrypt.email.example.file.MsgBlockParser.detectBlockNext(MsgBlockParser.kt:37)
        at com.flowcrypt.email.example.file.MsgBlockParser.detectBlocks(MsgBlockParser.kt:24)
        at com.flowcrypt.email.example.file.PgpMsg.detectBlockType(PgpMsg.kt:45)
        at com.flowcrypt.email.example.file.PgpKey.parseKeys(PgpKey.kt:87)
        at com.flowcrypt.email.example.file.PgpKey.parseKeysC(PgpKey.kt:128)
        at com.flowcrypt.email.example.file.PgpKey.parseKeysC$default(PgpKey.kt:127)
        at com.flowcrypt.email.jetpack.viewmodel.ContactsViewModel$getPgpContactInfoFromServer$2.invokeSuspend(ContactsViewModel.kt:305)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.UndispatchedCoroutine.afterResume(Builders.common.kt:214)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:113)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
java.lang.StringIndexOutOfBoundsException: length=49; index=50
        at java.lang.String.substring(String.java:1971)
        at com.flowcrypt.email.example.file.MsgBlockParser.detectBlockNext(MsgBlockParser.kt:37)
        at com.flowcrypt.email.example.file.MsgBlockParser.detectBlocks(MsgBlockParser.kt:24)
        at com.flowcrypt.email.example.file.PgpMsg.detectBlockType(PgpMsg.kt:45)
        at com.flowcrypt.email.example.file.PgpKey.parseKeys(PgpKey.kt:87)
        at com.flowcrypt.email.example.file.PgpKey.parseKeysC(PgpKey.kt:128)
        at com.flowcrypt.email.example.file.PgpKey.parseKeysC$default(PgpKey.kt:127)
        at com.flowcrypt.email.jetpack.viewmodel.ContactsViewModel$getPgpContactInfoFromServer$2.invokeSuspend(ContactsViewModel.kt:305)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.UndispatchedCoroutine.afterResume(Builders.common.kt:214)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:113)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

--- a/FlowCrypt/src/main/java/com/flowcrypt/email/core/msg/MsgBlockParser.kt
+++ b/FlowCrypt/src/main/java/com/flowcrypt/email/core/msg/MsgBlockParser.kt
@@ -21,23 +21,27 @@ object MsgBlockParser {
     val blocks = mutableListOf<MsgBlock>()
     var startAt = 0
     while (true) {
-      val continueAt = detectBlockNext(normalized, startAt, blocks)
+      val continueAt = detectNextBlock(normalized, startAt, blocks)
       if (startAt >= continueAt) return blocks
       startAt = continueAt
     }
   }
 
   @JvmStatic
-  private fun detectBlockNext(text: String, startAt: Int, blocks: MutableList<MsgBlock>): Int {
+  private fun detectNextBlock(text: String, startAt: Int, blocks: MutableList<MsgBlock>): Int {
     val initialBlockCount = blocks.size
     var continueAt = -1
     val beginIndex = text.indexOf(
-        PgpArmor.ARMOR_HEADER_DICT[MsgBlock.Type.UNKNOWN]!!.begin, startAt)
+        PgpArmor.ARMOR_HEADER_DICT[MsgBlock.Type.UNKNOWN]!!.begin, startAt
+    )
     if (beginIndex != -1) { // found
-      val potentialBeginHeader = text.substring(beginIndex, beginIndex + ARMOR_HEADER_MAX_LENGTH)
+      val potentialHeaderBegin = text.substring(
+          beginIndex,
+          (beginIndex + ARMOR_HEADER_MAX_LENGTH).coerceAtMost(text.length)
+      )
       for (blockHeaderKvp in PgpArmor.ARMOR_HEADER_DICT) {
         val blockHeaderDef = blockHeaderKvp.value
-        if (!blockHeaderDef.replace || potentialBeginHeader.indexOf(blockHeaderDef.begin) != 0) {
+        if (!blockHeaderDef.replace || potentialHeaderBegin.indexOf(blockHeaderDef.begin) != 0) {
           continue
         }
 

--- a/FlowCrypt/src/test/java/com/flowcrypt/email/core/msg/MsgBlockParserTest.kt
+++ b/FlowCrypt/src/test/java/com/flowcrypt/email/core/msg/MsgBlockParserTest.kt
@@ -15,6 +15,11 @@ import org.junit.Test
 
 class MsgBlockParserTest {
   @Test
+  fun testNoStringIndexOutOfBoundsExceptionInParser() {
+    checkForSinglePlaintextBlock("-----BEGIN FOO-----\n")
+  }
+
+  @Test
   fun testDetectBlocksDoesNotGetTrippedOnBlocksWithUnknownHeaders() {
     checkForSinglePlaintextBlock(
         "This text breaks email and Gmail web app.\n\n" +
@@ -273,9 +278,7 @@ Ek0f+P9DgunMb5OtkDwm6WWxpzV150LJcA==
     val blocks = MsgBlockParser.detectBlocks(input)
     assertEquals(1, blocks.size)
     assertTrue(blocks[0] is GenericMsgBlock)
-    assertEquals(
-        GenericMsgBlock(MsgBlock.Type.PLAIN_TEXT, input, true),
-        blocks[0] as GenericMsgBlock
-    )
+    val expectedBlock = GenericMsgBlock(MsgBlock.Type.PLAIN_TEXT, input.trimEnd(), true)
+    assertEquals(expectedBlock, blocks[0] as GenericMsgBlock)
   }
 }

--- a/script/ci-install-android-sdk.sh
+++ b/script/ci-install-android-sdk.sh
@@ -36,7 +36,7 @@ else
     (${ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager "emulator" | grep -v = || true)
     (${ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager "ndk;22.0.7026061" | grep -v = || true)
     (${ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager "cmake;3.10.2.4988404" | grep -v = || true)
-    (${ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager "system-images;android-30;google_apis;x86_64" | grep -v = || true)
+    (${ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager "system-images;android-29;google_apis;x86_64" | grep -v = || true)
 fi
 
 #Uncomment this for debug

--- a/script/ci-setup-and-run-emulator.sh
+++ b/script/ci-setup-and-run-emulator.sh
@@ -1,10 +1,7 @@
 #!/bin/bash
 
 "$ANDROID_SDK_ROOT/emulator/emulator" -accel-check
-echo -ne '\n' | avdmanager -v create avd --name ci-emulator --package "system-images;android-30;google_apis;x86_64" --device 'pixel_xl' --abi 'google_apis/x86_64'
-cat ~/.android/avd/ci-emulator.avd/config.ini
-echo "vm.heapSize=256"  >> ~/.android/avd/ci-emulator.avd/config.ini
-echo "hw.ramSize=3064"  >> ~/.android/avd/ci-emulator.avd/config.ini
+echo -ne '\n' | avdmanager -v create avd --name ci-emulator --package "system-images;android-29;google_apis;x86_64" --device 'pixel_xl' --abi 'google_apis/x86_64'
 cat ~/.android/avd/ci-emulator.avd/config.ini
 "$ANDROID_SDK_ROOT/emulator/emulator" -list-avds #debug
 "$ANDROID_SDK_ROOT/emulator/emulator" -avd ci-emulator -no-window -no-boot-anim -no-audio &

